### ISSUE TITLE
Issues with chunked transfer-encoding streams

### ISSUE THREAD

**Issue by amitga-rnd** [minio/minio-java#744]
Hi, I'm using the latest release of the minio-java client (5.0.6), and latest minio server.

I have a web app written in Java 8, using Jersey 2 as a Rest Server.
My app accepts uploads of files, and saves them to minio. Because I want to support large files, I use chunked transfer encoding.

When I upload to minio, and set the Content-Length parameter of `putObject` (which my client sends), the files (usually above 500 Mb) get corrupted (even though the upload was successful, when I download the file, and binary compare it to the original, I see its not the same).

If I **don't** set the Content-Length header parameter of `putObject`, the files are uploaded OK, and not corrupted. This is good, but if I limit my java memory usage (`-Xmx256m`) I get out of memory errors of the java heap.
Doing a little debugging, I see that its because around 550Mb are buffered to memory, when the minio client calculates the stream size (`getAvailableSize`).
Even if I increase the heap size to 1GB, I still get OOM (when uploading a 680 Mb file).
Only after increasing to about 2GB, it works.

Any help would be appreciated :-)

**Comment by kannappanr** [comment]
@amitga-rnd Thanks for reaching out to us. Based on what you are saying, I see that there are two issues

1. The Content-Length parameter in `putobject` if set, the data gets corrupted.
2. OOM error.

Can you provide a sample code to reproduce the first issue? 

We will also try to reproduce both errors and get back to you.

**Comment by amitga-rnd** [comment]
OK, I will try to provide some code, although its a bit complicated:
I forgot to say that the issue only happens when the `InputStream` given to `putObject` is a sort of network stream (not really seekable, I suppose). When I tried to reproduce using a `FileInputStream`, it doesn't happen. The same with streams based on memory (like `ByteArrayInputStream`), they all work fine.
I think the problem is because the minio client wraps the stream with a `BufferedInputStream` which does all kinds of things on the input stream, like seeking/setting position etc...
 
The stream received from the Jersey API is actually an `EntityInputStream` (even though its actually just a byte stream), which is some sort of network stream.

**Comment by amitga-rnd** [comment]
In order not to complicate the sample code with my web app, I have successfully reporoduced the problem reading a stream from URL.
I have downloaded the file locally too, and the test compares the contents of the local file, to the contents of the object received from minio after it was uploaded as a stream.
Here is the test I ran:
```
  @Test
    public void stream_test() throws Exception
    {
        URL movie = new URL("https://download.blender.org/peach/bigbuckbunny_movies/BigBuckBunny_320x180.mp4");
        URLConnection conn = movie.openConnection();
        int length = conn.getContentLength();

        String containerName = "test-container";

        if(!minioClient.bucketExists(containerName))
            minioClient.makeBucket(containerName);

        minioClient.putObject(containerName,"movie", conn.getInputStream() ,length,(String) null);

        // Removing the Content-Length parameter causes it to succeed, but uses up more memory, especially with larger files (~600 mb and up):
        // minioClient.putObject(containerName,"movie", conn.getInputStream(),(String) null);

        InputStream inputStream = minioClient.getObject(containerName,"movie");

        byte[] bytesFromObject = IOUtils.toByteArray(inputStream);
        byte[] bytesFromFile = FileUtils.readFileToByteArray(new File("/home/test/Downloads/BigBuckBunny_320x180.mp4"));

        Assert.assertTrue(Arrays.equals(bytesFromObject,bytesFromFile));
    }
```

**Comment by amitga-rnd** [comment]
Is there any update on this issue? 
Did my example help you reproduce the problem ?
Thanks in advance :-)

**Comment by sinhaashish** [comment]
> Is there any update on this issue?
> Did my example help you reproduce the problem ?
> Thanks in advance :-)

 I was able to replicate it . Its WIP as of now.

**Comment by nitisht** [comment]
fixed by #755
